File: server/test/colorTable.test.ts
import { describe, expect, it } from 'vitest';
import {
  ANALYSIS_SCHEME,
  COLOR_SCHEMES,
  DBZ_MIN,
  dbzToMmPerHour,
  decodePixel,
  intensityLabel,
  isSnowDistinguishable,
  legendFor,
  mmPerHourToDbz,
} from '../src/radar/colorTable.js';
import { RAIN_RAMPS, SNOW_RAMPS } from '../src/radar/colorTable.generated.js';

describe('tabla de colores', () => {
  it('publica los nueve esquemas del proveedor', () => {
    expect(COLOR_SCHEMES).toHaveLength(9);
    expect(COLOR_SCHEMES.map((s) => s.id)).toEqual([0, 1, 2, 3, 4, 5, 6, 7, 8]);
  });

  it('el esquema de análisis separa lluvia y nieve sin ambigüedad', () => {
    expect(isSnowDistinguishable(ANALYSIS_SCHEME)).toBe(true);
  });
});

describe('leyenda', () => {
  it('cada color proviene de la rampa real del esquema', () => {
    for (const scheme of COLOR_SCHEMES) {
      const ramp = RAIN_RAMPS[scheme.key];
      for (const stop of legendFor(scheme.key)) {
        const o = (stop.dbz - DBZ_MIN) * 4;
        const hex = (v: number): string => v.toString(16).padStart(2, '0');
        expect(stop.rain).toBe(`#${hex(ramp[o]!)}${hex(ramp[o + 1]!)}${hex(ramp[o + 2]!)}`);
      }
    }
  });

  it('los escalones suben en reflectividad e intensidad', () => {
    const stops = legendFor(ANALYSIS_SCHEME);
    expect(stops.length).toBeGreaterThan(5);
    for (let i = 1; i < stops.length; i++) {
      expect(stops[i]!.dbz).toBeGreaterThan(stops[i - 1]!.dbz);
      expect(stops[i]!.mmPerHour).toBeGreaterThan(stops[i - 1]!.mmPerHour);
    }
  });

  it('rotula los umbrales que ofrecen las alarmas', () => {
    const labelled = legendFor(ANALYSIS_SCHEME).filter((s) => s.label);
    expect(labelled.map((s) => s.dbz)).toEqual([20, 30, 40, 50]);
  });

  it('sólo incluye color de nieve si el esquema la distingue', () => {
    for (const scheme of COLOR_SCHEMES) {
      const distinguishable = isSnowDistinguishable(scheme.key);
      for (const stop of legendFor(scheme.key)) {
        expect(Boolean(stop.snow)).toBe(distinguishable);
      }
    }
  });
});

describe('decodePixel', () => {
  it('devuelve null para píxeles transparentes', () => {
    expect(decodePixel(0, 0, 0, 0)).toBeNull();
  });

  it('cada color de la rampa de lluvia vuelve a su propio color', () => {
    const ramp = RAIN_RAMPS[ANALYSIS_SCHEME];
    let checked = 0;
    for (let i = 0; i < 128; i++) {
      const o = i * 4;
      const a = ramp[o + 3]!;
      if (a === 0) continue;

      const decoded = decodePixel(ramp[o]!, ramp[o + 1]!, ramp[o + 2]!, a);
      expect(decoded).not.toBeNull();
      expect(decoded!.kind).toBe('rain');

      // La rampa repite color en algunos tramos altos; ante un empate se
      // devuelve el dBZ más bajo, que es el criterio conservador para alarmas.
      expect(decoded!.dbz).toBeLessThanOrEqual(DBZ_MIN + i);

      // Invariante fuerte: el nivel devuelto tiene exactamente ese color.
      const back = (decoded!.dbz - DBZ_MIN) * 4;
      expect([ramp[back], ramp[back + 1], ramp[back + 2], ramp[back + 3]]).toEqual([
        ramp[o],
        ramp[o + 1],
        ramp[o + 2],
        ramp[o + 3],
      ]);
      checked++;
    }
    expect(checked).toBeGreaterThan(90);
  });

  it('los niveles con color propio se recuperan de forma exacta', () => {
    const ramp = RAIN_RAMPS[ANALYSIS_SCHEME];
    // Umbrales que usa la configuración de sensibilidad.
    for (const dbz of [12, 20, 30, 40, 50]) {
      const o = (dbz - DBZ_MIN) * 4;
      const decoded = decodePixel(ramp[o]!, ramp[o + 1]!, ramp[o + 2]!, ramp[o + 3]!);
      expect(decoded!.dbz).toBe(dbz);
    }
  });

  it('clasifica como nieve los colores de la rampa de nieve', () => {
    const ramp = SNOW_RAMPS[ANALYSIS_SCHEME];
    for (let i = 40; i < 90; i++) {
      const o = i * 4;
      if ((ramp[o + 3] ?? 0) === 0) continue;
      const decoded = decodePixel(ramp[o]!, ramp[o + 1]!, ramp[o + 2]!, ramp[o + 3]!);
      expect(decoded?.kind).toBe('snow');
    }
  });

  it('tolera pequeñas desviaciones de color y descarta las grandes', () => {
    const ramp = RAIN_RAMPS[ANALYSIS_SCHEME];
    const o = 60 * 4; // dBZ 28
    const near = decodePixel(
      (ramp[o]! + 3) % 256,
      ramp[o + 1]!,
      ramp[o + 2]!,
      ramp[o + 3]!,
    );
    expect(near).not.toBeNull();

    // Un color que no se parece a nada de la paleta (magenta puro con alfa alto).
    expect(decodePixel(255, 0, 255, 255)).toBeNull();
  });

  it('cachea las búsquedas aproximadas devolviendo el mismo resultado', () => {
    const first = decodePixel(1, 165, 225, 255);
    const second = decodePixel(1, 165, 225, 255);
    expect(second).toEqual(first);
  });
});

describe('conversión de reflectividad', () => {
  it('la relación Marshall-Palmer es invertible', () => {
    for (const dbz of [10, 20, 30, 40, 50]) {
      expect(mmPerHourToDbz(dbzToMmPerHour(dbz))).toBeCloseTo(dbz, 6);
    }
  });

  it('crece con la reflectividad', () => {
    expect(dbzToMmPerHour(20)).toBeLessThan(dbzToMmPerHour(35));
    expect(dbzToMmPerHour(35)).toBeLessThan(dbzToMmPerHour(50));
  });

  it('sitúa 20 dBZ en torno a la lluvia débil', () => {
    const mmh = dbzToMmPerHour(20);
    expect(mmh).toBeGreaterThan(0.5);
    expect(mmh).toBeLessThan(1.5);
  });

  it('devuelve cero por debajo del mínimo de la escala', () => {
    expect(dbzToMmPerHour(DBZ_MIN)).toBe(0);
  });
});

describe('intensityLabel', () => {
  it('escala la etiqueta con la reflectividad', () => {
    expect(intensityLabel(5)).toBe('Sin precipitación');
    expect(intensityLabel(15)).toBe('Llovizna');
    expect(intensityLabel(25)).toBe('Lluvia débil');
    expect(intensityLabel(35)).toBe('Lluvia moderada');
    expect(intensityLabel(45)).toBe('Lluvia fuerte');
    expect(intensityLabel(55)).toBe('Tormenta');
  });
});

File: android/app/src/test/java/cat/plou/radar/ColorTableTest.kt
package cat.plou.radar

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class ColorTableTest {

    @Test
    fun `publica los nueve esquemas del proveedor`() {
        assertEquals(9, ColorScheme.entries.size)
        assertEquals(listOf(0, 1, 2, 3, 4, 5, 6, 7, 8), ColorScheme.entries.map { it.id })
    }

    @Test
    fun `el esquema de analisis separa lluvia y nieve sin ambiguedad`() {
        assertTrue(ColorDecoder(ANALYSIS_SCHEME).snowDistinguishable)
    }

    @Test
    fun `un color de la rampa se decodifica al dBZ que le corresponde`() {
        val decoder = ColorDecoder(ANALYSIS_SCHEME)
        val ramp = ANALYSIS_SCHEME.rain
        // Se comprueba la rampa entera: cada color debe devolver un dBZ que no
        // supere el suyo (los colores repetidos se resuelven al valor más bajo).
        for (i in 0 until (DBZ_MAX - DBZ_MIN + 1)) {
            val o = i * 4
            val a = ramp[o + 3].toInt() and 0xFF
            if (a == 0) continue
            val decoded = decoder.decode(
                ramp[o].toInt() and 0xFF,
                ramp[o + 1].toInt() and 0xFF,
                ramp[o + 2].toInt() and 0xFF,
                a,
            )
            assertNotNull("dBZ ${DBZ_MIN + i} sin decodificar", decoded)
            assertTrue(decoded!!.dbz <= DBZ_MIN + i)
            assertEquals(PrecipKind.RAIN, decoded.kind)
        }
    }

    @Test
    fun `los umbrales de las alarmas se decodifican de forma exacta`() {
        val decoder = ColorDecoder(ANALYSIS_SCHEME)
        for (level in Intensity.entries) {
            val i = level.dbz - DBZ_MIN
            val o = i * 4
            val decoded = decoder.decode(
                ANALYSIS_SCHEME.rain[o].toInt() and 0xFF,
                ANALYSIS_SCHEME.rain[o + 1].toInt() and 0xFF,
                ANALYSIS_SCHEME.rain[o + 2].toInt() and 0xFF,
                ANALYSIS_SCHEME.rain[o + 3].toInt() and 0xFF,
            )
            assertEquals("umbral ${level.name}", level.dbz, decoded?.dbz)
        }
    }

    @Test
    fun `un pixel transparente no es eco`() {
        assertNull(ColorDecoder().decode(0, 0, 0, 0))
    }

    @Test
    fun `un color lejano a la paleta se descarta`() {
        // Verde oscuro: a más de 40 por canal de cualquier entrada de la paleta.
        // (El verde puro sí existe en la rampa: corresponde a 75 dBZ.)
        assertNull(ColorDecoder(ColorScheme.UNIVERSAL_BLUE).decode(0, 128, 0, 255))
    }

    @Test
    fun `un color casi exacto cae en la entrada mas parecida`() {
        val decoder = ColorDecoder(ANALYSIS_SCHEME)
        val i = 30 - DBZ_MIN
        val o = i * 4
        val exacto = decoder.decode(
            ANALYSIS_SCHEME.rain[o].toInt() and 0xFF,
            ANALYSIS_SCHEME.rain[o + 1].toInt() and 0xFF,
            ANALYSIS_SCHEME.rain[o + 2].toInt() and 0xFF,
            255,
        )
        val desviado = decoder.decode(
            (ANALYSIS_SCHEME.rain[o].toInt() and 0xFF) + 2,
            (ANALYSIS_SCHEME.rain[o + 1].toInt() and 0xFF) - 1,
            (ANALYSIS_SCHEME.rain[o + 2].toInt() and 0xFF) + 1,
            255,
        )
        assertEquals(exacto?.dbz, desviado?.dbz)
    }

    @Test
    fun `la nieve se distingue de la lluvia`() {
        val decoder = ColorDecoder(ANALYSIS_SCHEME)
        val i = 30 - DBZ_MIN
        val o = i * 4
        val snow = decoder.decode(
            ANALYSIS_SCHEME.snow[o].toInt() and 0xFF,
            ANALYSIS_SCHEME.snow[o + 1].toInt() and 0xFF,
            ANALYSIS_SCHEME.snow[o + 2].toInt() and 0xFF,
            ANALYSIS_SCHEME.snow[o + 3].toInt() and 0xFF,
        )
        assertEquals(PrecipKind.SNOW, snow?.kind)
    }

    @Test
    fun `Marshall-Palmer ida y vuelta`() {
        for (mmh in listOf(0.1, 1.0, 5.0, 20.0, 100.0)) {
            val dbz = mmPerHourToDbz(mmh)
            assertEquals(mmh, dbzToMmPerHour(dbz), mmh * 1e-6)
        }
        assertEquals(0.0, dbzToMmPerHour(DBZ_MIN.toDouble()), 1e-9)
    }

    @Test
    fun `la leyenda sube en reflectividad e intensidad`() {
        val stops = legendFor(ANALYSIS_SCHEME)
        assertTrue(stops.size > 5)
        for (i in 1 until stops.size) {
            assertTrue(stops[i].dbz > stops[i - 1].dbz)
            assertTrue(stops[i].mmPerHour > stops[i - 1].mmPerHour)
        }
        assertEquals(listOf(20, 30, 40, 50), stops.filter { it.label != null }.map { it.dbz })
    }

    @Test
    fun `la etiqueta de intensidad sigue los umbrales`() {
        assertEquals("Sin precipitación", Intensity.labelFor(0.0))
        assertEquals("Llovizna", Intensity.labelFor(12.0))
        assertEquals("Lluvia débil", Intensity.labelFor(25.0))
        assertEquals("Tormenta", Intensity.labelFor(60.0))
    }
}
